Functions/xASL_quant_HadamardDecoding.m

```diff
@@ -0,0 +1,165 @@
+            DecodingMatrix = [1 -1  1 -1;
+                              1 -1 -1  1;
+                              1  1 -1 -1];
+        end
+        
+    elseif TimeEncodedMatrixSize == 8
+        
+        ASL_im = xASL_io_Nifti2Im(Encoded_ASL); % Load time-series nifti
+        
+        if ~isempty(DecodingMatrix_input) %if there's a Decoding Matrix from the data
+            DecodingMatrix = DecodingMatrix_input;
+        else
+            DecodingMatrix = [1 -1  1 -1  1 -1  1 -1;
+                              1 -1  1 -1 -1  1 -1  1;
+                              1 -1 -1  1 -1  1  1 -1;
+                              1 -1 -1  1  1 -1 -1  1;
+                              1  1 -1 -1  1  1 -1 -1;
+                              1  1 -1 -1 -1 -1  1  1;
+                              1  1  1  1 -1 -1 -1 -1];
+        end
+    end
+    
+    % #### For Hadamard Decoding Matrix ####
+elseif strcmp(TimeEncodedMatrixType,'Hadamard')
+    
+    
+    
+    if TimeEncodedMatrixSize == 4
+        
+        ASL_im = xASL_io_Nifti2Im(Encoded_ASL); % Load time-series nifti
+        
+        if ~isempty(DecodingMatrix_input) %if there's a Decoding Matrix from the data
+            DecodingMatrix = DecodingMatrix_input;
+        else
+            DecodingMatrix = [1 -1  1 -1;
+                              1  1 -1 -1;
+                              1 -1 -1  1];
+        end
+        
+    elseif TimeEncodedMatrixSize == 8
+        
+        ASL_im = xASL_io_Nifti2Im(Encoded_ASL); % Load time-series nifti
+        
+        if ~isempty(DecodingMatrix_input) %if there's a Decoding Matrix from the data
+            DecodingMatrix = DecodingMatrix_input;
+        else
+            DecodingMatrix = [1 -1 -1  1 -1  1  1 -1;
+                              1  1 -1 -1 -1 -1  1  1;
+                              1 -1  1 -1 -1  1 -1  1;
+                              1  1  1  1 -1 -1 -1 -1;
+                              1 -1 -1  1  1 -1 -1  1;
+                              1  1 -1 -1  1  1 -1 -1;
+                              1 -1  1 -1  1 -1  1 -1];
+        end
+    end
+    
+end
+
+%% step-1: Reorder data
+% At this point the data is organized like this (in terms of ASL4D.nii volumes):
+% PLD1/TE1,PLD1/TE2,PLD1/TE3,PLD1/TE4...PLD2/TE1,PLD2/TE2,PLD2/TE3... (PLDs first, TEs after)
+%
+% And for decoding we want
+% TE1/PLD1,TE1/PLD2,TE1/PLD3,TE1.PL4...TE2/PLD1,TE2/PLD2,TE2/PLD3,TE2/PLD4 (TEs first, PLDs after)
+
+numberTE = NumberEchoTimes;
+nDecodedTI = TimeEncodedMatrixSize-1;                   % Number of TIs is always MatrixSize -1
+nDecodedVol = numberTE * nDecodedTI;
+nEncodedVol = size(ASL_im, 4);
+EncodedDataSize = TimeEncodedMatrixSize * numberTE;     % Expected data size
+nRepetitions = nEncodedVol / EncodedDataSize;           % Calculating no. of acquisition repeats
+
+DecodedDataSize = nDecodedVol * nRepetitions;
+Decoded_ASL = zeros(size(ASL_im,1),size(ASL_im,2),size(ASL_im,3),DecodedDataSize);
+numberPLDs = int32(size(ASL_im,4)/numberTE);
```

Comment:
Why do we use `int32` here?

Reply:
It's an easy way to round the number, to be 100% certain that you multiply with an integer later on. I'd keep it like it is @BeatrizPadrela.